rmw_fastrtps_cpp/src/register_node.cpp

```diff
@@ -0,0 +1,217 @@
+      common_context.get()),
+    [&](CustomParticipantInfo * participant_info) {
+      if (RMW_RET_OK != rmw_fastrtps_shared_cpp::destroy_participant(participant_info)) {
+        fprintf(
+          stderr, "Failed to destroy participant after function: '"
+          RCUTILS_STRINGIFY(__function__) "' failed.\n");
+      }
+    });
+  if (!participant_info) {
+    return RMW_RET_BAD_ALLOC;
+  }
+
+  rmw_qos_profile_t qos = rmw_qos_profile_default;
+
+  qos.avoid_ros_namespace_conventions = false;  // Change this to true after testing
+  qos.history = RMW_QOS_POLICY_HISTORY_KEEP_LAST;
+  qos.depth = 1;
+  qos.durability = RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL;
+  qos.reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
+
+  std::unique_ptr<rmw_publisher_t, std::function<void(rmw_publisher_t *)>>
+  publisher(
+    rmw_fastrtps_cpp::create_publisher(
+      participant_info.get(),
+      rosidl_typesupport_cpp::get_message_type_support_handle<ParticipantEntitiesInfo>(),
+      "_participant_info",
+      &qos,
+      &publisher_options,
+      false,  // our fastrtps typesupport doesn't support keyed topics
+      true),
+    [&](rmw_publisher_t * pub) {
+      if (RMW_RET_OK != rmw_fastrtps_shared_cpp::destroy_publisher(
+        eprosima_fastrtps_identifier,
+        participant_info.get(),
+        pub))
+      {
+        fprintf(
+          stderr, "Failed to destroy publisher after function: '"
+          RCUTILS_STRINGIFY(__function__) "' failed.\n");
+      }
+    });
+  if (!publisher) {
+    return RMW_RET_BAD_ALLOC;
+  }
+
+  // If we would have support for keyed topics, this could be KEEP_LAST and depth 1.
+  qos.history = RMW_QOS_POLICY_HISTORY_KEEP_ALL;
+  std::unique_ptr<rmw_subscription_t, std::function<void(rmw_subscription_t *)>>
+  subscription(
+    rmw_fastrtps_cpp::create_subscription(
+      participant_info.get(),
+      rosidl_typesupport_cpp::get_message_type_support_handle<ParticipantEntitiesInfo>(),
+      "_participant_info",
+      &qos,
+      &subscription_options,
+      false,  // our fastrtps typesupport doesn't support keyed topics
+      true),
+    [&](rmw_subscription_t * sub) {
+      if (RMW_RET_OK != rmw_fastrtps_shared_cpp::destroy_subscription(
+        eprosima_fastrtps_identifier,
+        participant_info.get(),
+        sub))
+      {
+        fprintf(
+          stderr, "Failed to destroy subscription after function: '"
+          RCUTILS_STRINGIFY(__function__) "' failed.\n");
+      }
+    });
+  if (!subscription) {
+    return RMW_RET_BAD_ALLOC;
+  }
+
+  common_context->gid = rmw_fastrtps_shared_cpp::create_rmw_gid(
+    eprosima_fastrtps_identifier, participant_info->participant->getGuid());
+  common_context->pub = publisher.get();
+  common_context->sub = subscription.get();
+
+  context->impl->common = common_context.get();
+  context->impl->participant_info = participant_info.get();
+
+  if (RMW_RET_OK != rmw_fastrtps_cpp::run_listener_thread(context)) {
```

Comment:
Why not store and return the ret code from this function?

Reply:
agreed